### ISSUE TITLE
DynamicMap with hv.Overlay of a dynamic number of  hv.Curve  is broken

### ISSUE THREAD

**Issue by jameseflowers1000** [holoviz/holoviews#5681]
Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ]
Holoviews version: 1.15.0
Bokeh version: 2.4.3

Expect plot output to be consistent whether NCurvesStart in the code below is set to 1 or 3.  _It is not._
To see the bug:

- Run code as is with panel serve --autoreload --show.  
- Drag N slider and note both changing colors and amplitudes of sine waves as expected.

Then:

- Change NCurvesStart = 1 in the code and re-run the example.
- Use slider to change NCurves to 3.
- As before drag the N slider and note that the colors change but **NOT** the amplitudes.

```
import numpy as np
import panel as pn
import holoviews as hv
from random import random, choice

pn.extension('bokeh')

NCurvesStart = 3   # Everything works fine, change to 1 to see the bug

def sine(n, ncurves):
    curves = []
    colors = ['red', 'green', 'blue', 'orange', 'purple']
    for i in range(ncurves):
        random_amplitude = 1 + (random()-0.5)*0.3*i  # random amplitude
        frequency = 2.0
        xs = np.linspace(0.0, 1.0, n)
        ys = random_amplitude*np.sin((i+1)*frequency*xs)  # harmonic freq=(i+1)*frequency
        color = choice(colors) # choose a random color
        curve = hv.Curve((xs,ys), label=str(i),
                        kdims=['x'],vdims=['y'])
        curve.opts(color = color) # use the random color to show this callback is functional
        curves.append(curve)
    return hv.Overlay(curves)

ncurves = pn.widgets.IntSlider(name='NCurves', start=1, end=3, value=NCurvesStart).servable()
n = pn.widgets.IntSlider(name='N', start=100, end=500, value=500).servable()
col = pn.Column( hv.DynamicMap(pn.bind(sine, n, ncurves)),).servable()
```

Discussion:
It appears that this bug has been present for a long time.  I suspect it is related to [https://github.com/holoviz/holoviews/issues/4499]. 

Attempts made to work around:
Tried resetting cache in the DynamicMap - no change in behavior
Tried NdOverlay instead of Overlay - no change in behavior

Why is this important?:
Many applications need to present a varying number of curves based on user input or changing scenarios.   A DynamicMap is needed in order to maintain zoom and pan position, etc., in the presence of changing inputs.   This is a critical need and after two weeks of searching, no solution has been found.

**Comment by Hoxbro** [comment]
I only see the first curve not updating, and this seems to be because `random_amplitude` is always 1 for `i=0`.

Can you update Holoviews to version 1.15.4?

**Comment by jameseflowers1000** [comment]
Thanks so much.   It looks like it has indeed been fixed in 1.15.4.
